### ISSUE TITLE
Long press > "Share Image"  only provides a link to share. It does not share the actual image.

### ISSUE THREAD

**Issue by cyrusxs** [mozilla-mobile/fenix#12414]
## Steps to reproduce
1. Open image on Firefox Beta.
1. Long press the image > the following options show up:

* Share image
* Open image in new tab
* Save image
* Copy image location

![image](https://user-images.githubusercontent.com/11793648/86998785-f2590780-c165-11ea-81aa-823aeb824040.png)

1. Tap the "Share image" option.
1. The share sheet comes up, but it shows the url with a copy option, followed by all the apps I can share it to.

![image](https://user-images.githubusercontent.com/11793648/86998811-0270e700-c166-11ea-9d52-c416ab85bb24.png)

1. If I select WhatsApp, and choose a contact, it only shares the url to the image - it does not share the actual image.

![image](https://user-images.githubusercontent.com/11793648/86998822-0b61b880-c166-11ea-9b04-fb200e859cd1.png)

### Expected behavior
I expect that I should be able to share the actual image that I have long pressed on, rather than the link.

### Actual behavior
The url to the image is shared, not the actual image.

### Device information
* Android device:
Pixel 2 XL running Android 10
Build number: QQ3A.200605.001
* Fenix version: Firefox Beta v79.0.0-beta-4

**Comment by Racle** [comment]
+1

IMO there should be option to "share image" which will share image and "share image link" if you want only share link.

**Comment by SinTan1729** [comment]
+1
This used to work on fennec. I use facebook from browser and share memes to friends, so this is really annoying for me. Please fix it.

**Comment by ekager** [comment]
AC bug https://github.com/mozilla-mobile/android-components/issues/3648

**Comment by billybob884** [comment]
Just updated, still broken in latest build 81.1.1 #2015764547

**Comment by eliserichards** [comment]
It would be possible to have separate menu items for "share link to image" and "share image". What do you think @mozilla-mobile/ux?

**Comment by vesta0** [comment]
I think we should change the way "share image" works and just share the image (not the link). I doubt we need a separate "share link to image" menu item.

**Comment by violasong** [comment]
Yep, I agree that we don't need "Share Link to Image". We already have URL access via "Open image in new tab" and "Copy image location." I'm not sure we even need Copy Image Location. Do we have any data for these?

**Comment by apbitner** [comment]
Whatever decision is made here should probably also apply to this ticket needing UX feedback - #7458

**Comment by liuche** [comment]
So it sounds like the recommendation from @violasong is: change the functionality of "Share image" to (download and) pass the image to the Share intent?

**Comment by violasong** [comment]
Yes, assuming that is the same as what Vesta's comment is saying :)

**Comment by vesta0** [comment]
@liuche yes let's change the functionality of "share image" to download and pass the image (and not the image link) to the share intent. Also please note that this is not a UX ticket.

**Comment by billybob884** [comment]
> @liuche This is exactly how the "share image" used to already function before the last major update broke it.

**Comment by apbitner** [comment]
@violasong @vesta0 Based on this decision, can we close #7458?

**Comment by Mugurell** [comment]
With first downloading the image there is an interesting situation which arises when sharing bigger images / having slow network speed for which I'd need some UX feedback:
What I'm currently going with is a behavior similar to what happens when downloading an image:
- a notification is posted for the download with options to pause and cancel,
- when the download is complete instead of the "Download completed" dialog I'd show the share menu (the Android Sharesheet), again, just for the tab that from which the user pressed "Share image" and started the download.
- with the image now downloaded the share process

Does this behavior sounds ok?
I'm mainly interested in what should happen while the image is downloading. Is the notification enough? And is it ok to show the Android Sharesheet after the download is completed?

&nbsp;
Another scenario would be to have our own new sharesheet for this and let the user 
- press "Share image"
- select a share target in the new sharesheet

A download notification would be posted and only after the download is complete we'd actually start the share process.
This would be a more complicated approach and would lead to another interesting scenario: after selecting the share target the image would be downloaded in background, user would continue her work but then when the download is complete the share target app would appear on top of everything maybe in some cases unexpectedly so.

I'm also open to other suggestions.

**Comment by chronicweirdo** [comment]
Hello,

I think a good UX example is what is found in other browsers: since this feature is no longer available in Firefox on Android, I've been using the Chrome and Vivaldi browsers to share images on my device. On both browsers, the share image context menu will directly open options for the share intent, and once you choose an option will make the image available to the relevant application. As already mentioned, this is the way the share image functionality used to work for Firefox as well.

I don't think there is a need to download the image first. If the image is displayed on screen, that image has already been downloaded to your device and should be available somewhere, in some temp folder or cache. The share image functionality should use the image from that temp folder or cache.

**Comment by Mugurell** [comment]
The old Firefox would function the same - as in [it would also have to first download the image](https://dxr.mozilla.org/mozilla-esr68/rev/6863f516ba386cfadfd2f60ac45411dc7a0850cf/mobile/android/base/java/org/mozilla/gecko/widget/GeckoActionProvider.java#302).
User would press to share the image, it would be downloaded in the background with the user free to browse around and then the share menu would pop out when the image has finished downloading.
The image displayed in the webpage is only available in the cache for that page, it's not stored on disk / available to the app.
There is an open ticket for having GeckoView provide the image - https://bugzilla.mozilla.org/show_bug.cgi?id=1647568. 
@liuche @kbrosnan Maybe that should have a higher priority and block this ticket until it's available?

Even with that we can still get into edgecases where the image is not yet fully loaded, something like this happening for bigger images ([like the ones from here](https://visibleearth.nasa.gov/images/74393/july-blue-marble-next-generation-w-topography)) or when having a slow internet connection.

This can be seen now on Chrome which I think shares the image from cache.
Immediately after selecting the share option the share menu appears, user can choose an app to share to but then if the image is not yet available the share-to app opens without an image file/title for it actually shared.

**Comment by Mugurell** [comment]
If the image doesn't come from GeckoView and we must download it I thing another important detail is the location of the said image.
On Fennec these images were downloaded to and shared from a separate temporary directory and so the download part might seem transparent to the user.
Another option would be to treat them as normal downloads and have them appear in the downloaded files list and in the default download directory.

**Comment by lime124** [comment]
Chiming in while the rest of my team is part of the design sprint. I think it's fine to tackle these edge cases as @Mugurell suggests: 

This can be seen now on Chrome which I think shares the image from cache.
Immediately after selecting the share option the share menu appears, user can choose an app to share to but then if the image is not yet available the share-to app opens without an image file/title for it actually shared.

We shouldn't treat them as a normal download (that will be confusing to see it in the list since you didn't choose to download it) nor should we randomly pop up the share sheet after the image is finished downloading. at that point the person has probably moved on to something else and we're interrupting. i think doing your above suggestion is a fine compromise. thanks!

**Comment by tomyummmm** [comment]
Hi, its been quite a while and there hasn't been any updates on this situation. I'm still having to manually download and share images on whatsapp as a work around. Is there any chance of this being fixed?

**Comment by Racle** [comment]
> Hi, its been quite a while and there hasn't been any updates on this situation. I'm still having to manually download and share images on whatsapp as a work around. Is there any chance of this being fixed?

your app works like that?
If I download image with Firefox, it doesn't show at camera reel on Whatsapp and if I share that via gallery app, it shares that as file, not as image, so you cannot see the image, you have to download that..  

I usually copy URL, open chrome, paste URL there and share or download image from there (which is very annoying).

Firefox does something to image when it downloads that. Changes type from image to binary or something like that(??) and it doesn't show properly as image file on another apps.

**Comment by tomyummmm** [comment]
> your app works like that?
> If I download image with Firefox, it doesn't show at camera reel on Whatsapp and if I share that via gallery app, it shares that as file, not as image, so you cannot see the image, you have to download that..
> 
> I usually copy URL, open chrome, paste URL there and share or download image from there (which is very annoying).
> 
> Firefox does something to image when it downloads that. Changes type from image to binary or something like that(??) and it doesn't show properly as image file on another apps.

Whatsapp is able to see all photos i download in firefox, no problems. Still annoying to have to do this workaround. Is firefox unable to copy images to clipboard or is it unable to share images via cache to other apps like whatsapp?

**Comment by Racle** [comment]
> Whatsapp is able to see all photos i download in firefox, no problems. Still annoying to have to do this workaround. Is firefox unable to copy images to clipboard or is it unable to share images via cache to other apps like whatsapp?

I checked file MD5sum when downloaded image via chrome or firefox. No difference there. I can see file downloaded via chrome in whatsapp reel but not image downloaded from firefox.

But if I rename that Firefox file (like add 2 to end of filename before extension) it shows up in whatsapp. Even if I remove that 2 what I've added, it still works. But if I download same image again with firefox (without renaming), new image doesn't show up on whatsapp.

So file is same, filename seems to be somewhat different/broken for some reason (some character encoding? or something weird like that). Even tho chrome downloads new image with (1) added to that. 

This happens also with simple filenames like "sormi.jpg", so that excludes all special characters.

**Comment by nanowinner** [comment]
Hi, has there been any progress on this issue?

**Comment by leontancfa** [comment]
> Hi, has there been any progress on this issue?

You can vote the issue up

**Comment by apbitner** [comment]
@Mugurell do you need any additional feedback from UX beyond what @lime124 provided [above](https://github.com/mozilla-mobile/fenix/issues/12414#issuecomment-731347916)?

**Comment by Mugurell** [comment]
> @Mugurell do you need any additional feedback from UX beyond what @lime124 provided [above](https://github.com/mozilla-mobile/fenix/issues/12414#issuecomment-731347916)?

Thanks for pinging me on this.
Wanted to post a follow-up question since I was not really clear about what approach we should take but re-reading now again I see that we should go with:

- immediately after selecting the share option the share menu appears
- users can choose an app to share to
- the image is downloaded in the background in a private cache from which it will be removed after the share is complete

Accept that in some edge cases:
if the image is not yet available the share-to app opens without an image file/title for it actually shared.

Based on this indeed the ticket seems ready to be worked on.

**Comment by nanowinner** [comment]
> 
> 
> > @Mugurell do you need any additional feedback from UX beyond what @lime124 provided [above](https://github.com/mozilla-mobile/fenix/issues/12414#issuecomment-731347916)?
> 
> Thanks for pinging me on this.
> Wanted to post a follow-up question since I was not really clear about what approach we should take but re-reading now again I see that we should go with:
> 
>     * immediately after selecting the share option the share menu appears
> 
>     * users can choose an app to share to
> 
>     * the image is downloaded in the background in a private cache from which it will be removed after the share is complete
> 
>     * accept that in some edge cases - if the image is not yet available the share-to app opens without an image file/title for it actually shared.
> 
> 
> Based on this indeed the ticket seems ready to be worked on.

Great write up, btw! If you don't mind, could you share what those "edge cases" you refer to are?

**Comment by Mugurell** [comment]
> If you don't mind, could you share what those "edge cases" you refer to are?

Sharing the actual image entails asynchronicity - the image is downloaded in the background when the user selectsa the share option. And the flow will not wait for the image to be downloaded.
If the image is not downloaded when the user picks an app to share to that app will not receive an image.
This is happening with other apps also and it's a very interesting ux problem.
In most of the cases there should be enough time for the images to be downloaded but there could definitely be situations in which depending on the network / size of the image it will not be completely downloaded at the moment of actually sharing with an app.

**Comment by nanowinner** [comment]
![Screenshot_20210111_163607_com android chrome](https://user-images.githubusercontent.com/9197590/104202636-764cbe80-542b-11eb-8a6c-424a7f54847f.png)

I see. Chrome handles this interestingly by loading a thumbnail of the image intended above all options. One can see the spinning circle loading icon. 

Perhaps such a visual paired with a warning or a line of text informing the user to wait before sharing the image may slow down the quick thumbs of people just enough to reduce the chances of becoming one of thoseeedge cases.

**Comment by Mugurell** [comment]
> I see. Chrome handles this interestingly by loading a thumbnail of the image intended above all options. One can see the spinning circle loading icon.

I think Chrome is sharing the image from the browser cache. If the image has not yet been downloaded (and rendered on the screen) selecting the share option doesn't do anything. ([video showing Chrome's behavior](https://drive.google.com/file/d/16bdOtGC0NELEyCkVfIyzdglvwqPTMvem/view?usp=sharing))

We will be downloading it separately. Even after it is already rendered on the screen. We might change this in the future (@agi).
So like in the case with Chrome if there is not enough time for the image to be downloaded the share operation would fail. With maybe a different level of graciousness.

**Comment by Dunexus** [comment]
I don't know much about GV current priorities but isn't it possible to bump the priority of [BZ-1647568](https://bugzilla.mozilla.org/show_bug.cgi?id=1647568) and wait for it to be done to implement sharing from browser cache directly instead of implementing a custom download/cache mechanism and change the implementation later ?

**Comment by nanowinner** [comment]
This was a great and clear demonstration. 

> I think Chrome is sharing the image from the browser cache. If the image has not yet been downloaded (and rendered on the screen) selecting the share option doesn't do anything. ([video showing Chrome's behavior](https://drive.google.com/file/d/16bdOtGC0NELEyCkVfIyzdglvwqPTMvem/view?usp=sharing))

I will have to agree with @Dunexus. Since it's not too pressing of an issue, rather an inconvenience, perhaps doing it right the first time makes more sense in terms of not having to implement one method only to be replaced by another later. 

Keep up the great work, guys, and thanks for the time you took to read thisaand elaborat!

**Comment by BoyarkinNikita** [comment]
Firefox Nightly (210205), Android 11, OnePlus 8 Pro
Different behavior when sharing an image in Opera/Chrome and Firefox (wrong Intent or smthng)

[Firefox behavior](https://i.imgur.com/n8e0Zmk.png)

[Opera/Chrome behavior](https://i.imgur.com/VFedlnI.png)

[Image for test](https://meduza.io/impro/ZLl1xUZcZGGddYod_C0TkeJLNaUtBcOhv_aG6malvz0/fill/980/0/ce/1/aHR0cHM6Ly9tZWR1/emEuaW8vaW1hZ2Uv/YXR0YWNobWVudHMv/aW1hZ2VzLzAwNi80/NDMvNTg3L29yaWdp/bmFsL1h4NEk4N29D/Z2ROUkl6UXRtN1ls/WncuanBn.webp)

**Comment by Mugurell** [comment]
Thank you @BoyarkinNikita for the report.
I've opened a new ticket to investigate this - https://github.com/mozilla-mobile/fenix/issues/17898

**Comment by sflorean** [comment]
Verified as fixed on the latest Nightly build with LG G7 (Android 8.1.0).

**Comment by hinafu** [comment]
This is either not fixed or messed up again.

"Share Image" actually shares the URL to other apps. The Chrome app works as it should (actually Chrome displays 2 options "Share Image" and "Share Image URL").

**Comment by billybob884** [comment]
> This is either not fixed or messed up again.
> 
> "Share Image" actually shares the URL to other apps. The Chrome app works as it should (actually Chrome displays 2 options "Share Image" and "Share Image URL").

Agreed. Don't know why this was closed, it was never fixed

**Comment by Rodze** [comment]
> This is either not fixed or messed up again.
> 
> "Share Image" actually shares the URL to other apps. The Chrome app works as it should (actually Chrome displays 2 options "Share Image" and "Share Image URL").

It works for me on 20210305 nightly. The 2 options share image and share link are available

**Comment by Amejia481** [comment]
@hinafu and @billybob884 did you try it on Firefox [Nightly](https://play.google.com/store/apps/details?id=org.mozilla.fenix)?

**Comment by cousteaulecommandant** [comment]
I can confirm that it works on [Firefox Beta](https://play.google.com/store/apps/details?id=org.mozilla.firefox_beta) 87.0.0-beta.3, just the way it used to work before (it shares the image and the URL, not just the URL).
So the bug has been fixed and a new version is on the way; now all that's left to do is wait until mainstream [Firefox](https://play.google.com/store/apps/details?id=org.mozilla.firefox) gets updated to version 87 and gets the bug fix.

**Comment by tomyummmm** [comment]
I am on the beta and sharing the image to whatsapp causes an error message
that the file format is not supported. Is there any way to fix this?

**Comment by hinafu** [comment]
This feature did work when I tried it in Nightly, but the app was unusable in other aspects. By the second day it didn't even open.

I'll wait till this update reaches the "stable" release.

**Comment by Mugurell** [comment]
> I am on the beta and sharing the image to whatsapp causes an error message that the file format is not supported. Is there any way to fix this?

Please open a new ticket giving more information to help us reproduce the issue. Like:
- version of the Firefox used
- if possible the link to the image you are trying to share
- whether this issue reproduces when sharing the same from another apps

Thank you!

**Comment by cousteaulecommandant** [comment]
> I am on the beta and sharing the image to whatsapp causes an error message that the file format is not supported. Is there any way to fix this?

That might actually be WhatsApp's fault.  For example, I think WhatsApp cannot handle .webp images.  Did you try with other images?
(Maybe Firefox could convert images other than jpg/png/gif to a better supported format before sharing them, but that's a separate issue that should go on a new ticket)

**Comment by ErissonB** [comment]
I encountered the same problem and did not find the issue with the requested additional information, so I opened the issue #18547. @Mugurell 
I hope I've helped.

**Comment by orchidalloy** [comment]
I've also encountered the problem where images can't share to WhatsApp

**Comment by vsanjay85** [comment]
Firefox 90.1.2 and this issue still persists. Can't share the image to any app. Whatsapp straightaway says 'File format not supported' while all other apps simply show the URL to the image.

  What gives?

**Comment by tomyummmm** [comment]
Hi, when will this be implemented? Currently on Firefox 92.0.0 beta 8 and still unable to share images to whatsapp directly.

**Comment by mduchev** [comment]
@tomyummmm The original issue is fixed. Previously, there was no option to share the image itself, but only a url to it. Now you have that option and everything should be working fine. As far as I see from the comments above, the only issue that some people currently are experiencing is with WhatsApp. I suggest that you open another issue, since the issue created here is closed and with a good reason for that. Btw, I can share an image to WhatsApp without any problems at all, so the issue isn't as simple as "it doesn't work".

**Comment by leontancfa** [comment]
I can share to Whatsapp already. I see the issue resolved
